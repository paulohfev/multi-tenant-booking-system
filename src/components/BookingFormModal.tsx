import { useMemo, useState } from 'react';
import { useBookingsStore } from '@/store/useBookingsStore';
import { showToast } from '@/utils/showToast';
import { getCurrentDate, getCurrentTime } from '@/utils/date';
import { CreateBookingFormData } from '@/types/CreateBookingFormData';
import { Service } from '@/types/Service';
import InputField from './InputField';
import XIcon from './XIcon';

export type BookingFormModalProps = {
  service: Service;
  isOpen: boolean;
  onClose: () => void;
}

const BookingFormModal: React.FC<BookingFormModalProps> = ({
  service,
  isOpen,
  onClose,
}) => {
  const [formData, setFormData] = useState<CreateBookingFormData>({
    name: '',
    email: '',
    date: '',
    time: '',
  });

  const areAllFieldsFilled = useMemo(() => Object.values(formData).every((value) => value !== ''), [formData]);

  const { createBooking, isLoading } = useBookingsStore();

  const handleSubmit = async (e: React.FormEvent) => {
    e.preventDefault();
    try {
      await createBooking({
        ...formData,
        serviceName: service.name,
      });
      showToast('Booking created successfully!', 'success');
      onClose();
    } catch (error) {
      showToast(error as string, 'error');
    }
  };

  const handleChange = (
    e: React.ChangeEvent<HTMLInputElement | HTMLSelectElement>
  ) => {
    const { name, value } = e.target;
    setFormData((prev) => ({
      ...prev,
      [name]: value,
    }));
  };

  if (!isOpen) return null;

  const minTime = formData.date === getCurrentDate() ? getCurrentTime() : undefined;

  return (
    <div className="fixed inset-0 bg-gray-500/20 backdrop-brightness-50 flex items-center justify-center p-4">
      <div className="bg-white rounded-lg p-6 max-w-md w-full shadow-xl">
        <div className="flex justify-between items-center mb-4">
          <h2 className="text-2xl font-bold text-gray-900">
            Book {service.name}
          </h2>
          <button
            className="text-gray-500 hover:text-gray-700 cursor-pointer"
            onClick={onClose}
          >
            <XIcon classNameStyling="w-6 h-6" />
          </button>
        </div>

        <form onSubmit={handleSubmit} className="space-y-4">
          <InputField
            label="Name"
            name="name"
            type="text"
            value={formData.name}
            onChange={handleChange}
          />

          <InputField
            label="Email"
            name="email"
            type="email"
            value={formData.email}
            onChange={handleChange}
          />

          <InputField
            label="Date"
            name="date"
            type="date"
            value={formData.date}
            onChange={handleChange}
            min={getCurrentDate()}
          />

          <InputField
            label="Time"
            name="time"
            type="time"
            value={formData.time}
            onChange={handleChange}
            min={minTime}
          />

          <div className="mt-6">
            <button
              type="submit"
              disabled={isLoading || !areAllFieldsFilled}
              className="cursor-pointer w-full bg-blue-600 text-white py-2 px-4 rounded-md hover:bg-blue-700 transition-colors duration-200 disabled:bg-blue-400"
            >
              {isLoading ? 'Creating Booking...' : 'Confirm Booking'}
            </button>
          </div>
        </form>
      </div>
    </div>
  );
};

export default BookingFormModal; 