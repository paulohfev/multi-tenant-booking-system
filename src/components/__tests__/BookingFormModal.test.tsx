import { render } from '@testing-library/react'
import BookingFormModal from '../BookingFormModal'

const mockService = {
  id: '1',
  centerId: '1',
  name: 'Test Service',
  description: 'Test Description',
  duration: 60,
  price: 100,
}

describe('BookingFormModal', () => {
  it('should match snapshot when modal is open', () => {
    const { container } = render(<BookingFormModal service={mockService} isOpen={true} onClose={() => {}} />)
    expect(container).toMatchSnapshot()
  })

  it('should not render when modal is closed', () => {
    const { container } = render(<BookingFormModal service={mockService} isOpen={false} onClose={() => {}} />)
    expect(container).toMatchSnapshot()
  })
})
