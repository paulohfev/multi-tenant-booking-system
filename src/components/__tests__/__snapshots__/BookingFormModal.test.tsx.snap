// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookingFormModal should match snapshot when modal is open 1`] = `
<div>
  <div
    class="fixed inset-0 bg-gray-500/20 backdrop-brightness-50 flex items-center justify-center p-4"
  >
    <div
      class="bg-white rounded-lg p-6 max-w-md w-full shadow-xl"
    >
      <div
        class="flex justify-between items-center mb-4"
      >
        <h2
          class="text-2xl font-bold text-gray-900"
        >
          Book 
          Test Service
        </h2>
        <button
          class="text-gray-500 hover:text-gray-700 cursor-pointer"
        >
          <svg
            class="w-6 h-6"
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
          >
            <path
              d="M6 18L18 6M6 6l12 12"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </button>
      </div>
      <form
        class="space-y-4"
      >
        <div>
          <label
            class="block text-sm font-medium text-gray-700"
            for="name"
          >
            Name
          </label>
          <input
            class="mt-1 block w-full rounded-md border-gray-300 p-2 shadow-sm focus:border-blue-500 focus:ring-blue-500 text-black"
            id="name"
            name="name"
            required=""
            type="text"
            value=""
          />
        </div>
        <div>
          <label
            class="block text-sm font-medium text-gray-700"
            for="email"
          >
            Email
          </label>
          <input
            class="mt-1 block w-full rounded-md border-gray-300 p-2 shadow-sm focus:border-blue-500 focus:ring-blue-500 text-black"
            id="email"
            name="email"
            required=""
            type="email"
            value=""
          />
        </div>
        <div>
          <label
            class="block text-sm font-medium text-gray-700"
            for="date"
          >
            Date
          </label>
          <input
            class="mt-1 block w-full rounded-md border-gray-300 p-2 shadow-sm focus:border-blue-500 focus:ring-blue-500 text-black"
            id="date"
            min="2025-05-17"
            name="date"
            required=""
            type="date"
            value=""
          />
        </div>
        <div>
          <label
            class="block text-sm font-medium text-gray-700"
            for="time"
          >
            Time
          </label>
          <input
            class="mt-1 block w-full rounded-md border-gray-300 p-2 shadow-sm focus:border-blue-500 focus:ring-blue-500 text-black"
            id="time"
            name="time"
            required=""
            type="time"
            value=""
          />
        </div>
        <div
          class="mt-6"
        >
          <button
            class="cursor-pointer w-full bg-blue-600 text-white py-2 px-4 rounded-md hover:bg-blue-700 transition-colors duration-200 disabled:bg-blue-400"
            disabled=""
            type="submit"
          >
            Confirm Booking
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;

exports[`BookingFormModal should not render when modal is closed 1`] = `<div />`;
